File: agb/agb.py
# -*- coding: utf-8 -*-
"""
AGB represents the main client for interacting with the AGB cloud runtime
environment.
"""

import json
import os
from threading import Lock
from typing import Dict, List, Optional, Union

from agb.api.client import Client as mcp_client
from agb.api.models import (
    CreateSessionRequest,
    CreateSessionResponse,
    CreateMcpSessionRequestPersistenceDataList,
    GetSessionRequest,
    GetSessionResponse,
    ListSessionRequest,
    ListSessionResponse,
)
from agb.config import Config, load_config
from agb.model.response import DeleteResult, SessionResult, GetSessionResult, GetSessionData, SessionListResult
from agb.session import BaseSession, Session
from agb.session_params import CreateSessionParams
from agb.context import ContextService
from agb.logger import get_logger, log_operation_start, log_operation_success, log_warning

logger = get_logger(__name__)


class AGB:
    """
    AGB represents the main client for interacting with the AGB cloud runtime
    environment.
    """

    def __init__(self, api_key: str = "", cfg: Optional[Config] = None):
        """
        Initialize the AGB client.

        Args:
            api_key (str): API key for authentication. If not provided, it will be
                loaded from the AGB_API_KEY environment variable.
            cfg (Optional[Config]): Configuration object. If not provided, default
                configuration will be used.
        """
        if not api_key:
            api_key_env = os.getenv("AGB_API_KEY")
            if not api_key_env:
                raise ValueError(
                    "API key is required. Provide it as a parameter or set the "
                    "AGB_API_KEY environment variable"
                )
            api_key = api_key_env

        # Load configuration
        self.config = load_config(cfg)

        self.api_key = api_key
        self.endpoint = self.config.endpoint
        self.timeout_ms = self.config.timeout_ms

        # Initialize the HTTP API client with the complete config
        self.client = mcp_client(self.config)
        self._sessions: Dict[str, Session] = {}
        self._lock = Lock()

        # Initialize context service
        self.context = ContextService(self)

    def create(self, params: Optional[CreateSessionParams] = None) -> SessionResult:
        """
        Create a new session in the AGB cloud environment.

        Args:
            params (Optional[CreateSessionParams], optional): Parameters for
              creating the session.Defaults to None.

        Returns:
            SessionResult: Result containing the created session and request ID.
        """
        try:
            if params is None:
                params = CreateSessionParams()

            request = CreateSessionRequest(authorization=f"Bearer {self.api_key}")

            if params.image_id:
                request.image_id = params.image_id

             # Add labels if provided
            if params.labels:
                # Convert labels to JSON string
                request.labels = json.dumps(params.labels)

            # Flag to indicate if we need to wait for context synchronization
            needs_context_sync = False

            if params.context_syncs:
                persistence_data_list = []
                for context_sync in params.context_syncs:
                    if context_sync.policy:
                        policy_json = json.dumps(context_sync.policy.to_dict(), ensure_ascii=False)
                        persistence_data_list.append(CreateMcpSessionRequestPersistenceDataList(
                            context_id=context_sync.context_id,
                            path=context_sync.path,
                            policy=policy_json,
                        ))

                request.persistence_data_list = persistence_data_list
                needs_context_sync = len(persistence_data_list) > 0

            response: CreateSessionResponse = self.client.create_mcp_session(request)

            try:
                logger.debug("Response body:")
                logger.debug(response.to_dict())
            except Exception:
                logger.debug(f"Response: {response}")

            # Extract request ID
            request_id_attr = getattr(response, "request_id", "")
            request_id = request_id_attr or ""

            # Check if the session creation was successful
            if response.data and response.data.success is False:
                error_msg = response.data.err_msg
                if error_msg is None:
                    error_msg = "Unknown error"
                return SessionResult(
                    request_id=request_id,
                    success=False,
                    error_message=error_msg,
                )

            session_id = response.get_session_id()
            if not session_id:
                return SessionResult(
                    request_id=request_id,
                    success=False,
                    error_message=response.get_error_message(),
                )

            # ResourceUrl is optional in CreateMcpSession response
            resource_url = response.get_resource_url()

            logger.info(f"session_id = {session_id}")
            logger.info(f"resource_url = {resource_url}")

            # Create Session object
            session = Session(self, session_id)
            if resource_url is not None:
                session.resource_url = resource_url

            # Store image_id used for this session
            session.image_id = params.image_id or ""

            with self._lock:
                self._sessions[session_id] = session

            # If we have persistence data, wait for context synchronization
            if needs_context_sync:
                log_operation_start("Context synchronization", "Waiting for completion")

                # Wait for context synchronization to complete
                max_retries = 150  # Maximum number of retries
                retry_interval = 2  # Seconds to wait between retries

                import time
                for retry in range(max_retries):
                    # Get context status data
                    info_result = session.context.info()

                    # Check if all context items have status "Success" or "Failed"
                    all_completed = True
                    has_failure = False

                    for item in info_result.context_status_data:
                        logger.info(f"📁 Context {item.context_id} status: {item.status}, path: {item.path}")

                        if item.status != "Success" and item.status != "Failed":
                            all_completed = False
                            break

                        if item.status == "Failed":
                            has_failure = True
                            logger.error(f"❌ Context synchronization failed for {item.context_id}: {item.error_message}")

                    if all_completed or not info_result.context_status_data:
                        if has_failure:
                            log_warning("Context synchronization completed with failures")
                        else:
                            log_operation_success("Context synchronization")
                        break

                    logger.info(f"⏳ Waiting for context synchronization, attempt {retry+1}/{max_retries}")
                    time.sleep(retry_interval)

            # Return SessionResult with request ID
            return SessionResult(request_id=request_id, success=True, session=session)

        except Exception as e:
            logger.error(f"Error calling create_mcp_session: {e}")
            return SessionResult(
                request_id="",
                success=False,
                error_message=f"Failed to create session: {e}",
            )

    def list(
        self,
        labels: Optional[Dict[str, str]] = None,
        page: Optional[int] = None,
        limit: Optional[int] = None,
    ) -> SessionListResult:
        """
        Returns paginated list of session IDs filtered by labels.

        Args:
            labels (Optional[Dict[str, str]], optional): Labels to filter sessions.
                Defaults to None (empty dict).
            page (Optional[int], optional): Page number for pagination (starting from 1).
                Defaults to None (returns first page).
            limit (Optional[int], optional): Maximum number of items per page.
                Defaults to None (uses default of 10).

        Returns:
            SessionListResult: Paginated list of session IDs that match the labels,
                including request_id, success status, and pagination information.
        """
        try:
            # Set default values
            if labels is None:
                labels = {}
            if limit is None:
                limit = 10

            # Validate page number
            if page is not None and page < 1:
                return SessionListResult(
                    request_id="",
                    success=False,
                    error_message=f"Cannot reach page {page}: Page number must be >= 1",
                    session_ids=[],
                    next_token="",
                    max_results=limit,
                    total_count=0,
                )

            # Calculate next_token based on page number
            # Page 1 or None means no next_token (first page)
            # For page > 1, we need to make multiple requests to get to that page
            next_token = ""
            if page is not None and page > 1:
                # We need to fetch pages 1 through page-1 to get the next_token
                current_page = 1
                while current_page < page:
                    # Make API call to get next_token
                    labels_json = json.dumps(labels)
                    request = ListSessionRequest(
                        authorization=f"Bearer {self.api_key}",
                        labels=labels_json,
                        max_results=limit,
                    )
                    if next_token:
                        request.next_token = next_token

                    response = self.client.list_sessions(request)
                    request_id = getattr(response, "request_id", "") or ""

                    if not response.is_successful():
                        error_message = response.get_error_message() or "Unknown error"
                        return SessionListResult(
                            request_id=request_id,
                            success=False,
                            error_message=f"Cannot reach page {page}: {error_message}",
                            session_ids=[],
                            next_token="",
                            max_results=limit,
                            total_count=0,
                        )

                    next_token = response.get_next_token() or ""
                    if not next_token:
                        # No more pages available
                        return SessionListResult(
                            request_id=request_id,
                            success=False,
                            error_message=f"Cannot reach page {page}: No more pages available",
                            session_ids=[],
                            next_token="",
                            max_results=limit,
                            total_count=response.get_count() or 0,
                        )
                    current_page += 1

            # Make the actual request for the desired page
            labels_json = json.dumps(labels)
            request = ListSessionRequest(
                authorization=f"Bearer {self.api_key}",
                labels=labels_json,
                max_results=limit,
            )
            if next_token:
                request.next_token = next_token
                logger.debug(f"NextToken={request.next_token}")

            logger.info(f"Making list_sessions API call - Labels={labels_json}, MaxResults={limit}")

            # Make the API call
            response = self.client.list_sessions(request)

            # Extract request ID
            request_id = getattr(response, "request_id", "") or ""

            # Check for errors in the response
            if not response.is_successful():
                error_message = response.get_error_message() or "Unknown error"
                return SessionListResult(
                    request_id=request_id,
                    success=False,
                    error_message=f"Failed to list sessions: {error_message}",
                    session_ids=[],
                    next_token="",
                    max_results=limit,
                    total_count=0,
                )

            session_ids = []
            next_token = response.get_next_token() or ""
            max_results = response.get_max_results() or limit
            total_count = response.get_count() or 0

            # Extract session data
            session_data_list = response.get_session_data()

            # Process session data
            for session_data in session_data_list:
                if session_data.session_id:
                    session_ids.append(session_data.session_id)

            # Log API response with key details
            logger.info(f"ListSessions API response - RequestID: {request_id}, Success: True")
            logger.info(f"Total count: {total_count}, Returned count: {len(session_ids)}")
            logger.info(f"Has more: {'yes' if next_token else 'no'}")

            # Return SessionListResult with request ID and pagination info
            return SessionListResult(
                request_id=request_id,
                success=True,
                session_ids=session_ids,
                next_token=next_token,
                max_results=max_results,
                total_count=total_count,
            )

        except Exception as e:
            logger.error(f"Error calling list_sessions: {e}")
            return SessionListResult(
                request_id="",
                success=False,
                session_ids=[],
                error_message=f"Failed to list sessions: {e}",
            )

    def delete(self, session: BaseSession, sync_context: bool = False) -> DeleteResult:
        """
        Delete a session by session object.

        Args:
            session (BaseSession): The session to delete.
            sync_context (bool, optional): Whether to sync context before deletion. Defaults to False.

        Returns:
            DeleteResult: Result indicating success or failure and request ID.
        """
        try:
            # Delete the session and get the result
            delete_result = session.delete(sync_context=sync_context)

            with self._lock:
                self._sessions.pop(session.session_id, None)

            return delete_result

        except Exception as e:
            logger.error(f"Error calling delete_mcp_session: {e}")
            return DeleteResult(
                request_id="",
                success=False,
                error_message=f"Failed to delete session: {e}",
            )

    def get_session(self, session_id: str) -> GetSessionResult:
        """
        Get session information by session ID.

        Args:
            session_id (str): The ID of the session to retrieve.

        Returns:
            GetSessionResult: Result containing session information.
        """
        try:
            logger.info(f"Getting session information for session_id: {session_id}")

            request = GetSessionRequest(
                authorization=f"Bearer {self.api_key}",
                session_id=session_id
            )

            response = self.client.get_mcp_session(request)

            # Extract request ID from response
            request_id = getattr(response, "request_id", "") or ""

            try:
                response_body = json.dumps(
                    response.to_dict(), ensure_ascii=False, indent=2
                )
            except Exception:
                response_body = str(response)

            # Extract response information using your current architecture
            http_status_code = getattr(response, 'status_code', 0)
            code = getattr(response, 'code', "")
            success = response.is_successful() if hasattr(response, 'is_successful') else False
            message = response.get_error_message() if hasattr(response, 'get_error_message') else ""

                # Check for API-level errors
            if not success:
                return GetSessionResult(
                    request_id=request_id,
                    http_status_code=http_status_code,
                    code=code,
                    success=False,
                    data=None,
                    error_message=message or "Unknown error",
                )

            # Create GetSessionData from the API response using your architecture
            data = None
            if hasattr(response, 'data') and response.data:
                data = GetSessionData(
                    app_instance_id=response.data.app_instance_id or "",
                    resource_id=response.data.resource_id or "",
                    session_id=response.data.session_id or session_id,
                    success=True,
                    http_port=response.data.http_port or "",
                    network_interface_ip=response.data.network_interface_ip or "",
                    token=response.data.token or "",
                    vpc_resource=response.data.vpc_resource or False,
                    resource_url=response.data.resource_url or ""
                )

            # Log API response with key details
            key_fields = {}
            if data:
                key_fields["session_id"] = data.session_id
                key_fields["vpc_resource"] = "yes" if data.vpc_resource else "no"

            logger.info(f"GetSession API response - RequestID: {request_id}, Success: {success}")
            if key_fields:
                logger.debug(f"Key fields: {key_fields}")
            logger.debug(f"Full response: {response_body}")

            return GetSessionResult(
                request_id=request_id,
                http_status_code=http_status_code,
                code=code,
                success=success,
                data=data,
                error_message="",
            )

        except Exception as e:
            logger.error(f"Failed to parse response: {str(e)}")
            return GetSessionResult(
                request_id=request_id,
                success=False,
                error_message=f"Failed to parse response: {str(e)}",
            )
        except Exception as e:
            logger.error(f"Error calling GetSession: {e}")
            return GetSessionResult(
                request_id="",
                success=False,
                error_message=f"Failed to get session {session_id}: {e}",
            )

    def get(self, session_id: str) -> SessionResult:
        """
        Get a session by its ID.

        This method retrieves a session by calling the GetSession API
        and returns a SessionResult containing the Session object and request ID.

        Args:
            session_id (str): The ID of the session to retrieve.

        Returns:
            SessionResult: Result containing the Session instance, request ID, and success status.
        """
        # Validate input
        if not session_id or (isinstance(session_id, str) and not session_id.strip()):
            return SessionResult(
                request_id="",
                success=False,
                error_message="session_id is required",
            )

        # Call GetSession API
        get_result = self.get_session(session_id)

        # Check if the API call was successful
        if not get_result.success:
            error_msg = get_result.error_message or "Unknown error"
            return SessionResult(
                request_id=get_result.request_id,
                success=False,
                error_message=f"Failed to get session {session_id}: {error_msg}",
            )

        # Create the Session object
        session = Session(self, session_id)

        # Set session information from GetSession response
        if get_result.data:
            session.resource_url = get_result.data.resource_url or ""
            # Store additional session data - set attributes directly
            session.app_instance_id = get_result.data.app_instance_id or ""
            session.resource_id = get_result.data.resource_id or ""
            session.vpc_resource = get_result.data.vpc_resource or False
            session.network_interface_ip = get_result.data.network_interface_ip or ""
            session.http_port = get_result.data.http_port or ""
            session.token = get_result.data.token or ""

        logger.info(f"Successfully retrieved session: {session_id}")

        return SessionResult(
            request_id=get_result.request_id or "",
            success=True,
            session=session,
        )

File: typescript/src/agb.ts
import { Client } from "./api/client";
import {
    CreateSessionRequest,
    CreateMcpSessionRequestPersistenceDataList,
    GetSessionRequest,
    ListSessionRequest,
} from "./api/models";
import { Config, ConfigOptions, loadConfig } from "./config";
import {
    ContextSync,
    WhiteList,
    BWList,
    UploadPolicy,
    RecyclePolicy,
    SyncPolicy,
    newSyncPolicy,
    syncPolicyToDict,
} from "./context-sync";
import { ContextService } from "./context";
import { AuthenticationError } from "./exceptions";
import {
    logOperationStart,
    logOperationSuccess,
    logOperationError,
    logInfo,
    logWarn,
    logDebug,
} from "./logger";
import type {
    SessionResult,
    DeleteResult,
    GetSessionResult,
    GetSessionData,
    SessionListResult,
} from "./types/api-response";
import { Session } from "./session";
import type { CreateSessionParams } from "./session-params";
import { getSdkVersion, isReleaseVersion } from "./version";

const BROWSER_DATA_PATH = "/tmp/agb_browser_data";

/**
 * Options for initializing the AGB client.
 */
export interface AGBOptions {
    /** API key for authentication. If not provided, will use AGB_API_KEY environment variable. */
    apiKey?: string;
    /** Custom configuration. If not provided, uses environment-based configuration. */
    config?: ConfigOptions;
    /** Custom path to .env file. If not provided, searches upward from current directory. */
    envFile?: string;
}

/**
 * Main class for interacting with the AGB cloud runtime environment.
 * Use this class to create, list, and manage sessions.
 */
export class AGB {
    apiKey: string;
    config: Config;
    client: Client;
    /** Context service for managing persistent contexts. */
    context: ContextService;

    /**
     * Initialize the AGB client.
     *
     * @param options - Configuration options, or a string API key
     * @param options.apiKey - API key for authentication. If not provided, will use AGB_API_KEY environment variable.
     * @param options.config - Custom configuration object. If not provided, will use environment-based configuration.
     * @param options.envFile - Custom path to .env file. If not provided, will search upward from current directory.
     */
    constructor(options?: AGBOptions | string) {
        let apiKey: string;
        let cfgOptions: ConfigOptions | undefined;
        let envFile: string | undefined;

        if (typeof options === "string") {
            apiKey = options;
        } else {
            apiKey = options?.apiKey ?? "";
            cfgOptions = options?.config;
            envFile = options?.envFile;
        }

        if (!apiKey) {
            const envKey = process.env.AGB_API_KEY;
            if (!envKey) {
                throw new AuthenticationError(
                    "API key is required. Provide it as a parameter or set the AGB_API_KEY environment variable",
                );
            }
            apiKey = envKey;
        }

        this.config = loadConfig(cfgOptions, envFile);
        this.apiKey = apiKey;
        this.client = new Client(this.config);
        this.context = new ContextService(this);
    }

    /**
     * Serialize client configuration to a plain object.
     *
     * @returns Object with endpoint and timeoutMs
     */
    toJSON(): Record<string, unknown> {
        return {
            endpoint: this.config.endpoint,
            timeoutMs: this.config.timeoutMs,
        };
    }

    private normalizeContextSyncs(params: CreateSessionParams): ContextSync[] {
        const BROWSER_WHITELIST_PATHS: WhiteList[] = [
            { path: "/Local State", excludePaths: [] },
            { path: "/Default/Cookies", excludePaths: [] },
            { path: "/Default/Cookies-journal", excludePaths: [] },
        ];

        function createBrowserSyncPolicy(autoUpload: boolean): SyncPolicy {
            return newSyncPolicy({
                uploadPolicy: { autoUpload, uploadStrategy: "UploadBeforeResourceRelease" as any, uploadMode: "File" as any },
                bwList: { whiteLists: BROWSER_WHITELIST_PATHS } as BWList,
                recyclePolicy: { lifecycle: "Lifecycle_Forever" as any, paths: [""] },
            });
        }

        const syncs: ContextSync[] = [...(params.contextSync || [])];

        if (params.browserContext) {
            syncs.push(
                new ContextSync(
                    params.browserContext.contextId,
                    BROWSER_DATA_PATH,
                    createBrowserSyncPolicy(params.browserContext.autoUpload),
                ),
            );
        }

        return syncs;
    }

    private createPersistenceDataList(
        contextSyncs: ContextSync[],
    ): CreateMcpSessionRequestPersistenceDataList[] {
        const list: CreateMcpSessionRequestPersistenceDataList[] = [];
        for (const cs of contextSyncs) {
            if (cs.policy) {
                const policyJson = JSON.stringify(syncPolicyToDict(cs.policy));
                list.push(
                    new CreateMcpSessionRequestPersistenceDataList(
                        cs.contextId,
                        cs.path,
                        policyJson,
                    ),
                );
            }
        }
        return list;
    }

    /**
     * Create a new session in the AGB cloud environment.
     *
     * @param params - Parameters for creating the session (imageId is required)
     * @returns Promise resolving to SessionResult with success, session, requestId, and optional errorMessage
     *
     * @example
     * ```typescript
     * const agb = new AGB({ apiKey: 'your_api_key' });
     * const result = await agb.create({ imageId: 'agb-code-space-1' });
     * if (result.success && result.session) {
     *   await result.session.file.readFile('/etc/hostname');
     *   await agb.delete(result.session);
     * }
     * ```
     */
    async create(
        params?: CreateSessionParams | null,
    ): Promise<SessionResult> {
        try {
            if (!params) {
                const errorMessage = "params is required and cannot be null";
                logOperationError("AGB.create", errorMessage);
                return { requestId: "", success: false, errorMessage };
            }

            if (!params.imageId?.trim()) {
                const errorMessage =
                    "imageId is required and cannot be empty or null";
                logOperationError("AGB.create", errorMessage);
                return { requestId: "", success: false, errorMessage };
            }

            logOperationStart(
                "AGB.create",
                `ImageId=${params.imageId}`,
            );

            const request = new CreateSessionRequest(
                `Bearer ${this.apiKey}`,
                undefined,
                params.imageId,
            );

            if (params.policyId) {
                request.mcpPolicyId = params.policyId;
            }

            if (params.idleReleaseTimeout !== undefined) {
                request.timeout = params.idleReleaseTimeout;
            }

            if (params.labels) {
                request.labels = JSON.stringify(params.labels);
            }

            request.sdkStats = {
                source: "sdk",
                sdk_language: "typescript",
                sdk_version: getSdkVersion(),
                is_release: isReleaseVersion(),
            };

            let needsContextSync = false;
            const syncs = this.normalizeContextSyncs(params);
            const persistenceDataList = this.createPersistenceDataList(syncs);
            if (persistenceDataList.length > 0) {
                request.persistenceDataList = persistenceDataList;
                needsContextSync = true;
            }

            const response = await this.client.createMcpSession(request);
            const requestId = response.requestId ?? "";

            if (response.getData()?.success === false) {
                const errorMessage =
                    response.getData()?.errMsg ?? "Unknown error";
                logOperationError("AGB.create", errorMessage);
                return { requestId, success: false, errorMessage };
            }

            const sessionId = response.getSessionId();
            if (!sessionId) {
                const errorMessage =
                    response.getErrorMessage() ||
                    "Session ID not found in response";
                logOperationError("AGB.create", errorMessage);
                return { requestId, success: false, errorMessage };
            }

            const resourceUrl = response.getResourceUrl() ?? "";

            logInfo(`session_id = ${sessionId}`);
            logInfo(`resource_url = ${resourceUrl}`);

            const session = new Session(this, sessionId);
            session.resourceUrl = resourceUrl;
            session.imageId = params.imageId || "";

            const data = response.getData();
            if (data) {
                session.appInstanceId = data.appInstanceId ?? "";
                session.resourceId = data.resourceId ?? "";
            }

            if (needsContextSync) {
                await this.waitForContextSynchronization(session);
            }

            logOperationSuccess(
                "AGB.create",
                `SessionId=${sessionId}, RequestId=${requestId}`,
            );
            return { requestId, success: true, session };
        } catch (e: unknown) {
            const errMsg = e instanceof Error ? e.message : String(e);
            logOperationError("AGB.create", errMsg);
            return {
                requestId: "",
                success: false,
                errorMessage: `Failed to create session: ${errMsg}`,
            };
        }
    }

    private async waitForContextSynchronization(
        session: Session,
        maxRetries = 150,
        retryIntervalS = 2,
    ): Promise<void> {
        logOperationStart("Context synchronization", "Waiting for completion");
        const sleep = (ms: number) => new Promise((r) => setTimeout(r, ms));

        for (let retry = 0; retry < maxRetries; retry++) {
            const infoResult = await session.context.info();

            let allCompleted = true;
            let hasFailure = false;

            for (const item of infoResult.items) {
                logDebug(
                    `Context ${item.contextId} status: ${item.status}, path: ${item.path}`,
                );
                if (item.status !== "Success" && item.status !== "Failed") {
                    allCompleted = false;
                    break;
                }
                if (item.status === "Failed") {
                    hasFailure = true;
                }
            }

            if (allCompleted || infoResult.items.length === 0) {
                if (hasFailure) {
                    logWarn("Context synchronization completed with failures");
                } else {
                    logOperationSuccess("Context synchronization");
                }
                return;
            }

            logDebug(
                `Waiting for context synchronization, attempt ${retry + 1}/${maxRetries}`,
            );
            await sleep(retryIntervalS * 1000);
        }
    }

    /**
     * List sessions filtered by labels with pagination.
     *
     * @param labels - Optional key-value pairs to filter sessions
     * @param page - Page number (1-based). If provided, fetches that page.
     * @param limit - Maximum number of sessions per page (default 10)
     * @returns Promise resolving to SessionListResult with sessionIds, nextToken, totalCount, etc.
     */
    async list(
        labels?: Record<string, string>,
        page?: number,
        limit?: number,
    ): Promise<SessionListResult> {
        try {
            const effectiveLabels = labels ?? {};
            const effectiveLimit = limit ?? 10;

            logOperationStart(
                "AGB.list",
                `Page=${page ?? 1}, Limit=${effectiveLimit}`,
            );

            if (page !== undefined && page < 1) {
                const errorMessage = `Cannot reach page ${page}: Page number must be >= 1`;
                logOperationError("AGB.list", errorMessage);
                return {
                    requestId: "",
                    success: false,
                    errorMessage,
                    sessionIds: [],
                    nextToken: "",
                    maxResults: effectiveLimit,
                    totalCount: 0,
                };
            }

            let nextToken = "";
            if (page !== undefined && page > 1) {
                let currentPage = 1;
                while (currentPage < page) {
                    const req = new ListSessionRequest(
                        `Bearer ${this.apiKey}`,
                        effectiveLimit,
                        nextToken || undefined,
                        JSON.stringify(effectiveLabels),
                    );
                    const resp = await this.client.listSessions(req);
                    if (!resp.isSuccessful()) {
                        const errorMessage = `Cannot reach page ${page}: ${resp.getErrorMessage()}`;
                        return {
                            requestId: resp.requestId ?? "",
                            success: false,
                            errorMessage,
                            sessionIds: [],
                            nextToken: "",
                            maxResults: effectiveLimit,
                            totalCount: 0,
                        };
                    }
                    nextToken = resp.getNextToken() ?? "";
                    if (!nextToken) {
                        return {
                            requestId: resp.requestId ?? "",
                            success: false,
                            errorMessage: `Cannot reach page ${page}: No more pages available`,
                            sessionIds: [],
                            nextToken: "",
                            maxResults: effectiveLimit,
                            totalCount: resp.getCount() ?? 0,
                        };
                    }
                    currentPage++;
                }
            }

            const request = new ListSessionRequest(
                `Bearer ${this.apiKey}`,
                effectiveLimit,
                nextToken || undefined,
                JSON.stringify(effectiveLabels),
            );

            const response = await this.client.listSessions(request);
            const requestId = response.requestId ?? "";

            if (!response.isSuccessful()) {
                const errorMessage =
                    response.getErrorMessage() || "Unknown error";
                logOperationError("AGB.list", errorMessage);
                return {
                    requestId,
                    success: false,
                    errorMessage,
                    sessionIds: [],
                    nextToken: "",
                    maxResults: effectiveLimit,
                    totalCount: 0,
                };
            }

            const sessionIds: string[] = [];
            const sessionDataList = response.getSessionData();
            for (const sd of sessionDataList) {
                if (sd.sessionId) {
                    sessionIds.push(sd.sessionId);
                }
            }

            logOperationSuccess(
                "AGB.list",
                `RequestId=${requestId}, Count=${sessionIds.length}`,
            );

            return {
                requestId,
                success: true,
                sessionIds,
                nextToken: response.getNextToken() ?? "",
                maxResults: response.getMaxResults() ?? effectiveLimit,
                totalCount: response.getCount() ?? 0,
            };
        } catch (e: unknown) {
            const errMsg = e instanceof Error ? e.message : String(e);
            logOperationError("AGB.list", errMsg);
            return {
                requestId: "",
                success: false,
                errorMessage: `Failed to list sessions: ${errMsg}`,
                sessionIds: [],
                nextToken: "",
                maxResults: 0,
                totalCount: 0,
            };
        }
    }

    /**
     * Delete (release) a session.
     *
     * @param session - The session to delete
     * @param syncContext - If true, sync context before releasing (e.g. upload browser data)
     * @returns Promise resolving to DeleteResult with success and requestId
     */
    async delete(
        session: Session,
        syncContext = false,
    ): Promise<DeleteResult> {
        logOperationStart(
            "AGB.delete",
            `SessionId=${session.getSessionId()}, SyncContext=${syncContext}`,
        );
        try {
            const result = await session.delete(syncContext);
            if (result.success) {
                logOperationSuccess(
                    "AGB.delete",
                    `SessionId=${session.getSessionId()}`,
                );
            } else {
                logOperationError(
                    "AGB.delete",
                    result.errorMessage || "Unknown error",
                );
            }
            return result;
        } catch (e: unknown) {
            const errMsg = e instanceof Error ? e.message : String(e);
            logOperationError("AGB.delete", errMsg);
            return {
                requestId: "",
                success: false,
                errorMessage: `Failed to delete session: ${errMsg}`,
            };
        }
    }

    /**
     * Get session details by session ID (raw API result).
     *
     * @param sessionId - The session ID to look up
     * @returns Promise resolving to GetSessionResult with session data or error
     */
    async getSession(sessionId: string): Promise<GetSessionResult> {
        logOperationStart("AGB.getSession", `SessionId=${sessionId}`);
        try {
            const request = new GetSessionRequest(
                `Bearer ${this.apiKey}`,
                sessionId,
            );
            const response = await this.client.getMcpSession(request);
            const requestId = response.requestId ?? "";
            const success = response.isSuccessful();

            if (!success) {
                const errorMessage =
                    response.getErrorMessage() || "Unknown error";
                logWarn(`AGB.getSession: ${errorMessage}`);
                return {
                    requestId,
                    httpStatusCode: response.statusCode,
                    code: response.code ?? "",
                    success: false,
                    errorMessage,
                };
            }

            const data = response.getData();
            let sessionData: GetSessionData | undefined;
            if (data) {
                sessionData = {
                    appInstanceId: data.appInstanceId ?? "",
                    resourceId: data.resourceId ?? "",
                    sessionId: data.sessionId ?? sessionId,
                    success: true,
                    resourceUrl: data.resourceUrl ?? "",
                    status: data.status ?? "",
                };
            }

            logOperationSuccess(
                "AGB.getSession",
                `SessionId=${sessionId}, RequestId=${requestId}`,
            );

            return {
                requestId,
                httpStatusCode: response.statusCode,
                code: response.code ?? "",
                success: true,
                data: sessionData,
            };
        } catch (e: unknown) {
            const errMsg = e instanceof Error ? e.message : String(e);
            logOperationError("AGB.getSession", errMsg);
            return {
                requestId: "",
                httpStatusCode: 0,
                code: "",
                success: false,
                errorMessage: `Failed to get session ${sessionId}: ${errMsg}`,
            };
        }
    }

    /**
     * Get a Session instance by session ID.
     *
     * @param sessionId - The session ID
     * @returns Promise resolving to SessionResult with a Session object or error
     */
    async get(sessionId: string): Promise<SessionResult> {
        logOperationStart("AGB.get", `SessionId=${sessionId}`);

        if (!sessionId?.trim()) {
            const errorMessage = "sessionId is required";
            logOperationError("AGB.get", errorMessage);
            return { requestId: "", success: false, errorMessage, session: null };
        }

        const getResult = await this.getSession(sessionId);
        if (!getResult.success) {
            const errorMessage =
                getResult.errorMessage || "Unknown error";
            logOperationError("AGB.get", errorMessage);
            return {
                requestId: getResult.requestId ?? "",
                success: false,
                errorMessage: `Failed to get session ${sessionId}: ${errorMessage}`,
                session: null,
            };
        }

        const session = new Session(this, sessionId);
        if (getResult.data) {
            session.resourceUrl = getResult.data.resourceUrl || "";
            session.appInstanceId = getResult.data.appInstanceId || "";
            session.resourceId = getResult.data.resourceId || "";
        }

        logOperationSuccess(
            "AGB.get",
            `SessionId=${sessionId}, RequestId=${getResult.requestId}`,
        );

        return {
            requestId: getResult.requestId ?? "",
            success: true,
            session,
        };
    }
}
